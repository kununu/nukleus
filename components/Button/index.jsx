import React, {Component, PropTypes} from 'react';
import classNames from 'classnames';

import styles from './index.scss';

export default class Button extends Component {
  static propTypes = {
    customTheme: PropTypes.string,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    link: PropTypes.element,
    mobileFullWidth: PropTypes.bool,
    onClick: PropTypes.func,
    outline: PropTypes.bool,
    text: PropTypes.string,
    type: PropTypes.oneOf(['default', 'primary', 'secondary', 'info', 'danger', 'link', 'custom'])
  };

  static defaultProps = {
    text: '',
    type: 'primary'
  };

  onClick = e => {
    e.preventDefault();

    if (this.props.disabled) return;

    this.props.onClick();
  }

  render () {
    const {
      customTheme,
      disabled,
      fullWidth,
      link,
      mobileFullWidth,
      outline,
      text,
      type
    } = this.props;

    const classes = classNames({
      [styles.button]: true,
      [styles.fullWidth]: fullWidth,
      [styles.mobileFullWidth]: mobileFullWidth,
      [styles.outline]: outline,
      [type === 'custom' ? customTheme : styles[type]]: true
    });

    const props = {
      className: classes,
      disabled,
      onClick: this.props.onClick && this.onClick
    };

    if (link) {
      return (
        <div className={disabled ? styles.disabledLink : ''}>
          {React.cloneElement(link, props)}
        </div>
      );
    }

    return (
      <button
        {...props}>
        {text}
      </button>
    );
  }
}
