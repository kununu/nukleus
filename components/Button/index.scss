@import '~kununu-theme/scss/base/variables';

// Common styles
.button {
  border-radius: $border-radius-sm;
  font-weight: $font-weight-semibold;
  box-shadow: none;
  text-transform: uppercase;
  display: inline-block;
  text-align: center;
  vertical-align: middle;
  touch-action: manipulation;
  cursor: pointer;
  border: 1px solid transparent;
  white-space: nowrap;
  padding: 10px 12px;
  transition: background .15s;
  font-size: $font-size-base;

  &,
  &:hover,
  &:focus,
  &.focus,
  &:active,
  &.active {
    text-decoration: none;
  }

  &:focus,
  &.focus {
    outline: 5px auto -webkit-focus-ring-color;
    outline-offset: -2px;
  }

  &.disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &.focus,
    &:active,
    &.active {
      cursor: not-allowed;
      opacity: .65;
    }
  }

  &.fullWidth {
    width: 100%;
  }
}

@media (max-width: $screen-sm-min) {
  .button {
    white-space: normal;

    .mobileFullWidth {
      width: 100%;
    }
  }
}

@mixin button-default(
  $color,
  $background,
  $hover-background : lighten($background, 5%),
  $active-background: darken($background, 5%)
) {
  color: $color;
  background-color: $background;

  &:hover,
  &:focus,
  &.focus {
    color: $color;
    background-color: $hover-background;
  }

  &:active,
  &.active {
    color: $color;
    background-color: $active-background;
  }

  &.disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &.focus,
    &:active,
    &.active {
      color: $color;
      background-color: $background;
    }
  }
}

@mixin button-outline(
  $color,
  $background,
  $hover-color: darken($color, 5%),
  $active-color: lighten($color, 5%)
) {

  &.outline {
    color: $color;
    background-color: $background;
    border: 1px solid $color;

    &:hover,
    &:focus,
    &.focus {
      color: $hover-color;
      border-color: $hover-color;
    }

    &:active,
    &.active {
      color: $active-color;
      border-color: $active-color;
    }

    &.disabled,
    &[disabled] {
      &,
      &:hover,
      &:focus,
      &.focus,
      &:active,
      &.active {
        color: $color;
        background-color: $background;
      }
    }
  }
}

// Apply the mixin to the buttons
.default {
  @include button-default($btn-default-color, $btn-default-bg);
  @include button-outline($btn-default-color, $btn-info-bg);
}

.primary {
  @include button-default($btn-primary-color, $btn-primary-bg);
  @include button-outline($btn-primary-bg, $btn-primary-color);
}

.secondary {
  @include button-default($btn-secondary-color, $btn-secondary-bg);
  @include button-outline($btn-secondary-bg, $btn-secondary-color);
}

.info {
  @include button-default($btn-info-color, $btn-info-bg);
  @include button-outline($btn-info-color, $btn-info-bg);
}

.danger {
  @include button-default($btn-danger-color, $btn-danger-bg);
  @include button-outline($btn-danger-bg, $btn-danger-color);
}

.link {
  @include button-default($btn-secondary-bg, $btn-info-bg);
  @include button-outline($btn-secondary-bg, $btn-info-bg);

  &:hover {
    text-decoration: underline;
    border-color: transparent;
  }

  &.disabled,
  &[disabled] {
    &,
    &:hover,
    &:focus,
    &.focus,
    &:active,
    &.active {
      text-decoration: none;
    }
  }
}

.disabledLink {
  cursor: not-allowed;

  a {
    pointer-events: none;
  }
}
