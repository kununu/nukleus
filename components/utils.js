/* eslint-disable import/prefer-default-export */

/**
 * Replaces New Lines and multiple spaces and returns the new value
 *
 * @return {string} [updated value]
 */
export function sanitizeWhitespace (value) {
  return value.replace(/\r\n|\r|\n| +/g, ' ');
}
