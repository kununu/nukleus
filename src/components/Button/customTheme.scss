.button {
  color: white;
  border-radius: 5px;
  border: 0;
  text-transform: uppercase;
  font-weight: bold;
  padding: 10px 20px;
  box-shadow: 0px 5px 40px -11px rgba(0,0,0,0.76);
}

.primary {
  background-color: pink;
}