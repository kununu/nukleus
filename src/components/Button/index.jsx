import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';

import ThemeContext from 'utils/themeContext';
import themeable from 'utils/theming';

import styles from './index.scss';

export default class Button extends React.Component {
  static propTypes = {
    customTheme: PropTypes.string,
    disabled: PropTypes.bool,
    fullWidth: PropTypes.bool,
    htmlType: PropTypes.oneOf([
      'button',
      'submit',
      'reset',
    ]),
    link: PropTypes.element,
    mobileFullWidth: PropTypes.bool,
    onClick: PropTypes.func,
    outline: PropTypes.bool,
    text: PropTypes.oneOfType([
      PropTypes.element,
      PropTypes.string,
    ]),
    title: PropTypes.string,
    type: PropTypes.oneOf(['default', 'primary', 'secondary', 'info', 'danger', 'link', 'custom']),
    id: PropTypes.string,
  };

  static defaultProps = {
    customTheme: '',
    disabled: false,
    fullWidth: false,
    htmlType: 'submit',
    link: null,
    mobileFullWidth: false,
    onClick: null,
    outline: false,
    text: '',
    title: null,
    type: 'primary',
    id: null,
  };

  onClick = (e) => {
    const {
      disabled,
      onClick,
    } = this.props;

    e.preventDefault();

    if (disabled) return;

    onClick();
  }

  render () {
    const {
      customTheme,
      disabled,
      fullWidth,
      link,
      mobileFullWidth,
      onClick,
      outline,
      text,
      htmlType,
      title,
      type,
      id,
    } = this.props;

    const classes = classNames({
      button: true,
      fullWidth: fullWidth,
      mobileFullWidth: mobileFullWidth,
      outline: outline,
      [type === 'custom' ? customTheme : type]: true,
    }).split(' ')

    return (
      <ThemeContext.Consumer>
        {(context) => {
          const theme = themeable({...styles, ...context})

          const props = {
            className: theme(...classes),
            disabled,
            title,
            onClick: onClick && this.onClick,
            id,
          };

          if (link) {
            return (
              <div className={disabled ? styles.disabledLink : ''}>
                {React.cloneElement(link, props)}
              </div>
            );
          }

          return (
            <button // eslint-disable-line react/button-has-type
              {...props}
              type={htmlType}
            >
              {text}
            </button>
          );
        }}
      </ThemeContext.Consumer>
    );
  }
}
