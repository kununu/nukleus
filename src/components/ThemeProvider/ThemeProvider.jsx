import React from 'react';

import ThemeContext from 'utils/themeContext';

function ThemeProvider (props) {
  const {
    children,
    theme,
  } = props;

  return <ThemeContext.Provider value={theme}>{children}</ThemeContext.Provider>;
}

export default ThemeProvider;