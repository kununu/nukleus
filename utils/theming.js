const theming = theme => (...names) => names
  .map(name => theme[name] || name)
  .filter(x => x)
  .join(' ');

export default theming;
