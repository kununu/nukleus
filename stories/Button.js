import React from 'react';
import {storiesOf} from '@storybook/react';

import Button from '../src/components/Button';

storiesOf('Test', module)
  .add('button', () => (
    <Button text="hi" />
  ));
