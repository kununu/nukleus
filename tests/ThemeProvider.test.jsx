import React from 'react';
import renderer from 'react-test-renderer';
import Button from 'Button'; // eslint-disable-line import/no-unresolved
import * as themeable from '../utils/theming';
import ThemeProvider from 'ThemeProvider';


test('Renders correctly themed nukleus component', () => {
  const component = renderer.create(<ThemeProvider theme={{button: 'customTheme---button-14235'}}>
    <div>
      <Button text="Test" />
    </div>
  </ThemeProvider>);

  const tree = component.toJSON();

  expect(tree).toMatchSnapshot();
});
