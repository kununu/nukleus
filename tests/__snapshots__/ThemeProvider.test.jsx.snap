// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly themed nukleus component 1`] = `
<div>
  <button
    className="customTheme---button-14235 primary"
    disabled={false}
    id={null}
    onClick={null}
    title={null}
    type="submit"
  >
    Test
  </button>
</div>
`;

exports[`Renders theme provider without crashing 1`] = `
<div>
  <button
    className="button primary"
    disabled={false}
    id={null}
    onClick={null}
    title={null}
    type="submit"
  >
    Test
  </button>
</div>
`;
